server-spi/src/main/java/org/keycloak/vault/VaultTranscriber.java

```diff
@@ -0,0 +1,112 @@
+     * <p/>
+     * The returned {@link VaultCharSecret} extends {@link AutoCloseable} and it is strongly recommended that it is used in
+     * try-with-resources blocks to ensure the raw secret is overridden (destroyed) when the calling code is finished using
+     * it.
+     *
+     * @param value a {@link String} that might be a vault expression containing a vault entry key.
+     * @return a {@link VaultRawSecret} representing the secret that was read from the vault. If the specified value is not
+     *         a vault expression then the returned secret is the value itself encoded as a {@link VaultRawSecret}.
+     */
+    VaultCharSecret getCharSecret(final String value);
+
+    /**
+     * Obtains the secret represented as a {@link String} from the vault that matches the entry in the specified value.
+     * The value must follow the format {@code ${vault.<KEY>}} where {@code <KEY>} identifies the entry in the vault. If
+     * the value doesn't follow the vault expression format, it is assumed to be the secret itself.
+     * <p/>
+     * To avoid unnecessary manipulation of the value string it is recommended that the {@link #isVaultExpression(String)}
+     * method is used to check if the value is a vault entry string and then call this method only when that is true.
+     * <p/>
+     * <b>WARNING:</b> this method, unlike the other variations, doesn't return an {@link AutoCloseable} secret because
+     * strings cannot be overridden (destroyed) so there's no point in having a special type to enclose the secret
+     * string for usage in try-with-resources blocks. This means the secret might be visible in memory for as long as the
+     * string object exists. For this reason, the constructed string is returned as a {@link WeakReference} to convey that
+     * the secret string shouldn't be held longer than needed and to make sure it is available for the GC as soon as possible.
+     *
+     * @param value a {@link String} that might be a vault expression containing a vault entry key.
+     * @return an {@link Optional} holding the weak reference to the constructed secret string. If the specified value is not
+     *         a vault expression then the returned secret is the value itself enclosed in a {@link WeakReference}.
+     */
+    Optional<WeakReference<String>> getStringSecret(final String value);
+
+    /**
+     * Determines if the specified value is a vault expression referencing a vault entry key or not. A vault entry key is
+     * specified using the format {@code ${vault.<KEY>}}. For example, the string {@code ${vault.smtp_secret}} identifies
+     * an entry in the vault whose key is {@code smtp_secret}.
+     *
+     * @param value the {@code String} to be checked.
+     * @return {@code true} if the given value is a vault expression following the {@code ${vault.<KEY>}} format;
+     *          {@code false} otherwise.
+     */
+    boolean isVaultExpression(final String value);
```

Comment:
Is this method really needed?

Reply:
I would probably vote for removing this. In the state how it is now one need to double-treat the same else branch (basically two very similar ifs). One `isVaultExpression` and the second `Optional.isPresent()`. I don't think this will have a lot of performance issues with encoding all values to *Secret since this will not be used so often.

Reply:
Removed from interface.